pkg/profiler/profile_writer.go

```diff
@@ -59,29 +60,34 @@ func (fw *FileProfileWriter) Write(_ context.Context, labels model.LabelSet, pro
 // RemoteProfileWriter is a profile writer that writes profiles to a remote profile store.
 type RemoteProfileWriter struct {
 	profileStoreClient profilestorepb.ProfileStoreServiceClient
+	// pool of gzip encoders helps to reduce GC pressure.
+	pool sync.Pool
 }
 
 // NewRemoteProfileWriter creates a new RemoteProfileWriter.
 func NewRemoteProfileWriter(profileStoreClient profilestorepb.ProfileStoreServiceClient) *RemoteProfileWriter {
 	return &RemoteProfileWriter{
 		profileStoreClient: profileStoreClient,
+		pool: sync.Pool{New: func() interface{} {
+			z, _ := gzip.NewWriterLevel(nil, gzip.BestSpeed)
+			return z
+		}},
 	}
 }
 
 // Write sends the profile using the designated write client.
 func (rw *RemoteProfileWriter) Write(ctx context.Context, labels model.LabelSet, prof *profile.Profile) error {
 	buf := bytes.NewBuffer(nil)
-	zw, err := gzip.NewWriterLevel(buf, gzip.StatelessCompression)
-	if err != nil {
-		return err
-	}
-	if err = prof.WriteUncompressed(zw); err != nil {
-		zw.Close()
+	zw, _ := rw.pool.Get().(*gzip.Writer)
+	zw.Reset(buf)
+	if err := prof.WriteUncompressed(zw); err != nil {
+		rw.pool.Put(zw)
```

Comment:
Don't we need to close the `zw` regardless?

Reply:
Since `zw` gets reset before it's used again, a `Close` call is redundant (it writes a gzip footer). I will add it for symmetry since that code path won't be taken often anyways.

Reply:
Thanks for the explanation ❤️